wgsl/index.bs

```diff
@@ -5339,6 +5339,101 @@ Issue: [[#uniform-control-flow]] needs to state whether all invocations being di
   </xmp>
 </div>
 
+### Control Flow Interruption ### {#control-flow-interruption-section}
+
+A <dfn>control flow interruption</dfn> is any one of the following statements:
+* a [=break=] statement
+* a [=continue=] statement
+* a [=fallthrough=] statement
+* a [=return=] statement
+* a [=discard=] statement
+* a [=compound statement=] containing a [=control flow interruption=].
+
+A control flow interruption must appear last in its immediately enclosing compound statement.
+
+Note: Statements that immediately follow a control flow interruption,
+but within the same enclosing compound statement, can never be executed.
+These may indicate a programming error.
+
+[=Switch=], [=if=], [=loop=], and [=for=] statements *contain* [=compound statements=],
+but are not themselves compound statements.
+(That is, they do not match the `compound_statement` grammar rule.)
+They are never classified as control flow interruptions.
+
+<div class='example wgsl' heading='A simple invalid control flow interruption'>
+  <xmp highlight='rust'>
+   fn simple() -> i32 {
+     var a: i32;
+     return 0;  // Error: A control flow interruption, not appearing last.
+     a = 1;     // This is never executed.
+     return 2;
+   }
+  </xmp>
+</div>
+
+<div class='example wgsl' heading='An invalid nested control flow interruption'>
+  <xmp highlight='rust'>
+   fn nested() -> i32 {
+     var a: i32;
+     {             // The start of a compound statement.
+       a = 2;
+       return 1;   // Valid: A control flow interruption appearing last in its
+                   // enclosing compound statement.
+
+     }             // Error: The compound statement is a control flow interruption,
+                   // not appearing last.
+     a = 1;        // This is never executed.
+     return 2;
+   }
+  </xmp>
+</div>
+
+<div class='example wgsl' heading='An if is not a control flow interruption'>
+  <xmp highlight='rust'>
+   fn if_example() {
+     var a: i32 = 0;
+     loop {
+       // This if statement contains a control flow interruption.
+       // But a whole if statements is never a control flow interruption.
+       if (a == 5) {
+         break; // A control flow interruption.
+       }
+       a = a + 1;
+     }
+   }
+  </xmp>
+</div>
+
+<div class='example wgsl' heading='An invalid non-nested continue'>
+  <xmp highlight='rust'>
+   fn simple_continue() {
+     var a: i32;
+     loop {
+       if (a == 5) { break; }  // Valid
+       continue;               // Error: Invalid control flow interruption.
+       a = a + 1;
+     }                         // Valid
+   }
+  </xmp>
+</div>
+
+<div class='example wgsl' heading='An invalid non-nested continue with continuing statement'>
+  <xmp highlight='rust'>
+   fn continue_with_continuing() {
+     var a: i32;
+     loop {
+       if (a == 5) { break; }
+       continue;   // Error: Invalid control flow interruption.
```

Comment:
I feel that we should allow this. There isn't any instruction here that will never execute. This code is technically fully semantically healthy, just a tiny bit redundant.

Reply:
Yeah, that's why I included this example.  It's a harmless edge case.  I initially sided on keeping the rule as simple as possible.

The other weird one is:


       loop {
         //// blah blah blah
         continue;  // nothing after it.
       }

It's semantically healthy too, and should probably be allowed.

Reply:
continue right at the end *is* allowed. never mind.